In [ ]:
!pip install -q pyomo

     |████████████████████████████████| 9.6 MB 8.5 MB/s 
     |████████████████████████████████| 49 kB 6.0 MB/s 


In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
# create a model
model = ConcreteModel()

In [ ]:
N = 14

In [ ]:
obj_coef_c = np.array([-4.2, 37.75, 3.5, 10.05, -8.25, 40.85, 0.05, -11.12, 0.65, 20.725, -102.65, 0.6, -9.45, -25.625])

In [ ]:
constr1_coef_p = np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,1])
constr2_coef_q = np.array([2.8, -0.6725, 3.6, 0, -2.02, 1.01, -4.305, 0, 1, -2.09, 0, -2.124, -3.402, 0])
constr1_rhs_b1 = 343.48
constr2_rhs_b2 = 66.2

In [ ]:
lower_bound = np.array([2, -3, -np.inf, 1, 7, 3, -9, 0, 1, -25.245, -np.inf, -6, -30.5, -10.8])
upper_bound = np.array([np.inf, 6, 2.5, 3.4, np.inf, 6, 4.1, np.inf, 30, 11, 0, 0, 95, 13.4])

In [ ]:
col_indices = np.arange(N)
print(col_indices)


[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13]


In [ ]:
model.x = Var(col_indices)

In [ ]:
model.constraints = ConstraintList()

In [ ]:
model.constraints.add(sum(constr1_coef_p[j]*model.x[j] for j in col_indices) <= constr1_rhs_b1)

In [ ]:
model.constraints.add(sum(constr2_coef_q[j]*model.x[j] for j in col_indices) >= constr2_rhs_b2)


In [ ]:
model.constraints.add(sum(constr2_coef_q[j]*model.x[j] for j in col_indices) <= constr2_rhs_b2)


In [ ]:
for j in col_indices:
  model.x[j].setlb(lower_bound[j])
  model.x[j].setub(upper_bound[j])

In [ ]:
model.objective = Objective(expr = sum(obj_coef_c[j]*model.x[j] for j in col_indices), sense=minimize)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {1, 2, 3}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   14 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13}

1 Var Declarations
    x : Size=14, Index=x_index
        Key : Lower   : Value : Upper : Fixed : Stale : Domain
          0 :     2.0 :  None :  None : False :  True :  Reals
          1 :    -3.0 :  None :   6.0 : False :  True :  Reals
          2 :    None :  None :   2.5 : False :  True :  Reals
          3 :     1.0 :  None :   3.4 : False :  True :  Reals
          4 :     7.0 :  None :  None : False :  True :  Reals
          5 :     3.0 :  None :   6.0 : False :  True :  Reals
          6 :    -9.0 :  None :   4.1 : False :  True :  Reals
          7 :     0.0 :  None :  None : False :  True :  Reals
          8 :     1.0 : 

In [ ]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155680 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -14246.25704375
  Upper bound: -14246.25704375
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 15
  Number of nonzeros: 35
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.008849859237670898
# ----------------------------------------------------------
#   Solution Information
# --------------

In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  -14246.257043750004

Decision Variables
x[ 0 ] =  2039.2619375
x[ 1 ] =  -3.0
x[ 2 ] =  -1623.2369375
x[ 3 ] =  1.0
x[ 4 ] =  7.0
x[ 5 ] =  3.0
x[ 6 ] =  -9.0
x[ 7 ] =  0.0
x[ 8 ] =  1.0
x[ 9 ] =  -25.245
x[ 10 ] =  0.0
x[ 11 ] =  -6.0
x[ 12 ] =  -30.5
x[ 13 ] =  -10.8

Constraints
constraints : Size=3
    Key : Lower : Body               : Upper
      1 :  None : 343.47999999999985 : 343.48
      2 :  66.2 :  66.19999999999925 :   None
      3 :  None :  66.19999999999925 :   66.2


Answer:

1.Optimal values of decision variables:=
$x_0 =  2039.2619375 \\
x_1 =  -3.0 \\
x_2 =  -1623.2369375 \\
x_3 =  1.0 \\
x_4=  7.0 \\ 
x_5 =  3.0 \\ 
x_6 =  -9.0 \\ 
x_7 =  0.0 \\
x_8 =  1.0 \\
x_9 =  -25.245 \\
x_10 =  0.0 \\
x_11 =  -6.0 \\
x_12 =  -30.5 \\
x_13 =  -10.8  \\ $

2.Optimal value of the objective function:=-14246.257043750004
 

 

